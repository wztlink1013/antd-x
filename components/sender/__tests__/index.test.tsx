import React from 'react';

import { fireEvent, render } from '@testing-library/react';
import mountTest from '../../../tests/shared/mountTest';
import rtlTest from '../../../tests/shared/rtlTest';
import Sender from '../index';

describe('Sender Component', () => {
  mountTest(() => <Sender />);

  rtlTest(() => <Sender />);

  beforeAll(() => {
    jest.useFakeTimers();
  });

  afterAll(() => {
    jest.useRealTimers();
  });

  it('loading state', () => {
    const { asFragment } = render(<Sender loading />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('action as ReactNode', () => {
    const { container } = render(<Sender actions={<div className="bamboo" />} />);
    expect(container.querySelector('.bamboo')).toBeTruthy();
  });

  it('custom action button', () => {
    const onSubmit = jest.fn();
    const { container, getByText } = render(
      <Sender
        actions={(_, info) => {
          const { SendButton, ClearButton } = info.components;
          return (
            <div className="bamboo">
              <SendButton onClick={onSubmit} disabled={false}>
                SendPrompt
              </SendButton>
              <ClearButton disabled />
            </div>
          );
        }}
      />,
    );

    // check children render
    const sendButton = getByText('SendPrompt');
    expect(sendButton).toBeInTheDocument();

    const clearButton = container.querySelector('.bamboo button[disabled]');
    expect(clearButton).toBeInTheDocument();

    // check custom onClick
    fireEvent.click(sendButton);
    expect(onSubmit).toHaveBeenCalled();
  });

  it('onSubmit', () => {
    const onSubmit = jest.fn();
    const { container } = render(<Sender value="bamboo" onSubmit={onSubmit} />);
    fireEvent.click(container.querySelector('button')!);
    expect(onSubmit).toHaveBeenCalledWith('bamboo');
  });

  it('onCancel', () => {
    const onCancel = jest.fn();
    const { container } = render(<Sender loading onCancel={onCancel} />);
    fireEvent.click(container.querySelector('button')!);
    expect(onCancel).toHaveBeenCalled();
  });

  it('onClear', () => {
    const onChange = jest.fn();
    const { container } = render(
      <Sender
        onChange={onChange}
        actions={(_, { components: { ClearButton } }) => <ClearButton />}
      />,
    );

    fireEvent.change(container.querySelector('textarea')!, { target: { value: 'bamboo' } });
    expect(onChange).toHaveBeenCalledWith('bamboo', {});

    fireEvent.click(container.querySelector('button')!);
    expect(onChange).toHaveBeenCalledWith('', undefined);
  });

  describe('submitType', () => {
    it('default', () => {
      const onSubmit = jest.fn();
      const onPressKey = jest.fn();
      const { container } = render(
        <Sender value="bamboo" onSubmit={onSubmit} onKeyPress={onPressKey} />,
      );
      fireEvent.keyDown(container.querySelector('textarea')!, { key: 'Enter', shiftKey: false });
      expect(onSubmit).toHaveBeenCalledWith('bamboo');
      expect(onPressKey).toHaveBeenCalled();
    });

    it('shiftEnter', () => {
      const onSubmit = jest.fn();
      const { container } = render(
        <Sender value="bamboo" onSubmit={onSubmit} submitType="shiftEnter" />,
      );
      fireEvent.keyDown(container.querySelector('textarea')!, { key: 'Enter', shiftKey: true });
      expect(onSubmit).toHaveBeenCalledWith('bamboo');
    });
  });

  it('Sender.Header can be focus', () => {
    const { container } = render(
      <Sender
        header={
          <Sender.Header open>
            <input className="target" />
          </Sender.Header>
        }
      />,
    );

    const inputEle = container.querySelector<HTMLInputElement>('.target')!;
    inputEle.focus();
    expect(document.activeElement).toEqual(inputEle);

    // Click on the header
    fireEvent.mouseDown(container.querySelector('.ant-sender-header')!);
    expect(document.activeElement).toEqual(inputEle);

    // Click on the content
    fireEvent.mouseDown(container.querySelector('.ant-sender-content')!);
    expect(document.activeElement).toEqual(container.querySelector('textarea'));
  });

  it('readOnly', () => {
    const { container } = render(<Sender readOnly />);
    expect(container.querySelector('textarea')).toHaveAttribute('readonly');
  });

  describe('paste events', () => {
    it('onPaste callback', () => {
      const onPaste = jest.fn();
      const { container } = render(<Sender onPaste={onPaste} />);

      const textarea = container.querySelector('textarea')!;
      fireEvent.paste(textarea);
      expect(onPaste).toHaveBeenCalled();
      const eventArg = onPaste.mock.calls[0][0];
      expect(eventArg.type).toBe('paste');
      expect(eventArg.target).toBe(textarea);
    });

    it('onPasteFile callback with files', () => {
      const onPasteFile = jest.fn();
      const { container } = render(<Sender onPasteFile={onPasteFile} />);

      const file = new File(['test'], 'test.png', { type: 'image/png' });
      const fileList = {
        0: file,
        length: 1,
        item: (idx: number) => (idx === 0 ? file : null),
      };

      const textarea = container.querySelector('textarea')!;
      fireEvent.paste(textarea, {
        clipboardData: {
          files: fileList,
          getData: () => '',
        },
      });

      expect(onPasteFile).toHaveBeenCalledWith(file, fileList);
    });

    it('should not trigger onPasteFile when no files', () => {
      const onPasteFile = jest.fn();
      const { container } = render(<Sender onPasteFile={onPasteFile} />);

      const textarea = container.querySelector('textarea')!;
      fireEvent.paste(textarea, {
        clipboardData: {
          files: { length: 0 },
          getData: () => '',
        },
      });

      expect(onPasteFile).not.toHaveBeenCalled();
    });

    it('should handle multiple files paste', () => {
      const onPasteFile = jest.fn();
      const { container } = render(<Sender onPasteFile={onPasteFile} />);

      const file1 = new File(['test1'], 'test1.png', { type: 'image/png' });
      const file2 = new File(['test2'], 'test2.jpg', { type: 'image/jpeg' });
      const fileList = {
        0: file1,
        1: file2,
        length: 2,
        item: (idx: number) => (idx === 0 ? file1 : idx === 1 ? file2 : null),
      };

      const textarea = container.querySelector('textarea')!;
      fireEvent.paste(textarea, {
        clipboardData: {
          files: fileList,
          getData: () => '',
        },
      });

      expect(onPasteFile).toHaveBeenCalledWith(file1, fileList);
    });
  });
});
