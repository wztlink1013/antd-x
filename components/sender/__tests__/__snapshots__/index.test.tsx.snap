// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sender Component loading state 1`] = `
<DocumentFragment>
  <div
    class="ant-sender"
  >
    <div
      class="ant-sender-content"
    >
      <textarea
        class="ant-input ant-input-borderless ant-sender-input"
        style="overflow-y: hidden; resize: none;"
      />
      <div
        class="ant-sender-actions-list"
      >
        <div
          class="ant-sender-actions-list-presets ant-flex"
        >
          <button
            class="ant-btn ant-btn-circle ant-btn-text ant-btn-color-primary ant-btn-variant-text ant-sender-actions-btn ant-sender-actions-btn-loading-button"
            type="button"
          >
            <svg
              class="ant-sender-actions-btn-loading-icon"
              color="currentColor"
              viewBox="0 0 1000 1000"
              xmlns="http://www.w3.org/2000/svg"
            >
              <title>
                Stop Loading
              </title>
              <rect
                fill="currentColor"
                height="250"
                rx="24"
                ry="24"
                width="250"
                x="375"
                y="375"
              />
              <circle
                cx="500"
                cy="500"
                fill="none"
                opacity="0.45"
                r="450"
                stroke="currentColor"
                stroke-width="100"
              />
              <circle
                cx="500"
                cy="500"
                fill="none"
                r="450"
                stroke="currentColor"
                stroke-dasharray="600 9999999"
                stroke-width="100"
              >
                <animatetransform
                  attributeName="transform"
                  dur="1s"
                  from="0 500 500"
                  repeatCount="indefinite"
                  to="360 500 500"
                  type="rotate"
                />
              </circle>
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Sender Component rtl render component should be rendered correctly in RTL direction 1`] = `
<div
  class="ant-sender ant-sender-rtl"
>
  <div
    class="ant-sender-content"
  >
    <textarea
      class="ant-input ant-input-borderless ant-sender-input"
      style="overflow-y: hidden; resize: none;"
    />
    <div
      class="ant-sender-actions-list"
    >
      <div
        class="ant-sender-actions-list-presets ant-flex ant-flex-rtl"
      >
        <button
          class="ant-btn ant-btn-circle ant-btn-primary ant-btn-color-primary ant-btn-variant-solid ant-btn-icon-only ant-btn-rtl ant-sender-actions-btn ant-sender-actions-btn-disabled"
          type="button"
        >
          <span
            class="ant-btn-icon"
          >
            <span
              aria-label="arrow-up"
              class="anticon anticon-arrow-up"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="arrow-up"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M868 545.5L536.1 163a31.96 31.96 0 00-48.3 0L156 545.5a7.97 7.97 0 006 13.2h81c4.6 0 9-2 12.1-5.5L474 300.9V864c0 4.4 3.6 8 8 8h60c4.4 0 8-3.6 8-8V300.9l218.9 252.3c3 3.5 7.4 5.5 12.1 5.5h81c6.8 0 10.5-8 6-13.2z"
                />
              </svg>
            </span>
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
